### PR TITLE
Support batch updates in the worker sender 

### PR DESCRIPTION
**Pull request by anoadragon453**
We added the ability to batch updates but didn't handle this correctly in the communication between Synapse master and worker processes. Marking updates as batched included setting their `token`s to None which the relevant code choked on. This PR change this to handle them correctly.

Fixes https://github.com/matrix-org/synapse/issues/4705

### CODE REVIEW

File: synapse/replication/tcp/protocol.py [hunk 1]
```diff
@@ -478,10 +478,30 @@ def subscribe_to_stream(self, stream_name, token):
...
+                    batch_updates.append(update)
+                    if token and not token > current_token:
+                        # This batch is older than current_token, dismiss
+                        batch_updates = []
+                        continue
+                    if token:
+                        # Send all updates that are part of this batch with the
+                        # found token
+                        for update in batch_updates:
+                            self.send_command(RdataCommand(stream_name, token, update))
```
Comment (erikjohnston) [inline synapse/replication/tcp/protocol.py]: Don't we need to set `batch_updates = []`?

File: synapse/replication/tcp/protocol.py [hunk 2]
```diff
@@ -478,10 +478,30 @@ def subscribe_to_stream(self, stream_name, token):
...
+                else:
+                    # Only send updates newer than the current token
+                    if token > current_token:
+                        self.send_command(RdataCommand(stream_name, token, update))
```
Comment (erikjohnston) [inline synapse/replication/tcp/protocol.py]: This is a little nit picky, but I think this would read better if written like:

```python
if token is None:
    batch_updates.append(update)
elif current_token < token:
    batch_updates = []
else:
    batch_updates.append(update)
    for update in batch_updates:
        self.send_command(RdataCommand(stream_name, token, update))
    batch_updates = []
```

(with comments)

**Comment by erikjohnston** [comment]
It still feels quite a lot of nested ifs that is quite hard to follow, imo

**Comment by anoadragon453** [comment]
@erikjohnston Clearer now?
